src/Infrastructure.Dapper/AdminConsole/Helpers/BulkResourceCreationService.cs

```diff
@@ -0,0 +1,129 @@
+﻿using System.Data;
+using Bit.Core.Entities;
+using Bit.Core.Enums;
+using Microsoft.Data.SqlClient;
+
+namespace Bit.Infrastructure.Dapper.AdminConsole.Helpers;
+
+public static class BulkResourceCreationService
+{
+    private const string _defaultErrorMessage = "Must have at least one record for bulk creation.";
+    public static async Task CreateCollectionsUsersAsync(SqlConnection connection, SqlTransaction transaction, IEnumerable<CollectionUser> collectionUsers, string errorMessage = _defaultErrorMessage)
+    {
+        using var bulkCopy = new SqlBulkCopy(connection, SqlBulkCopyOptions.KeepIdentity, transaction);
+        bulkCopy.DestinationTableName = "[dbo].[CollectionUser]";
+        var dataTable = BuildCollectionsUsersTable(bulkCopy, collectionUsers, errorMessage);
+        await bulkCopy.WriteToServerAsync(dataTable);
+    }
+
+    private static DataTable BuildCollectionsUsersTable(SqlBulkCopy bulkCopy, IEnumerable<CollectionUser> collectionUsers, string errorMessage)
+    {
+        var collectionUser = collectionUsers.First();
+
+        if (collectionUser == null)
+        {
+            throw new ApplicationException(errorMessage);
+        }
+
+        var table = new DataTable("CollectionUserDataTable");
+
+        var collectionIdColumn = new DataColumn(nameof(collectionUser.CollectionId), typeof(Guid));
+        table.Columns.Add(collectionIdColumn);
+
+        var orgUserIdColumn = new DataColumn(nameof(collectionUser.OrganizationUserId), typeof(Guid));
+        table.Columns.Add(orgUserIdColumn);
+
+        var readOnlyColumn = new DataColumn(nameof(collectionUser.ReadOnly), typeof(bool));
+        table.Columns.Add(readOnlyColumn);
+
+        var hidePasswordsColumn = new DataColumn(nameof(collectionUser.HidePasswords), typeof(bool));
+        table.Columns.Add(hidePasswordsColumn);
+
+        var manageColumn = new DataColumn(nameof(collectionUser.Manage), typeof(bool));
+        table.Columns.Add(manageColumn);
```

Comment:
I suggest using `GetType()` like the CreateCollectionsAsync method does, so it automatically tracks any future change to the types used by the entity object.

Reply:
I agree, it doesn't work as nicely for nullable. That's fine.